File: src/python/day7/puzzle1.py
from .equation import equation

def isSolveable(eq):
    if len(eq.terms) == 1:
        return eq.target == eq.terms[-1]
    last = eq.terms[-1]

    if eq.target % last == 0 and isSolveable(equation(eq.target//last, eq.terms[:-1])):
        return True

    return isSolveable(equation(eq.target-last, eq.terms[:-1]))

if __name__ == "__main__":
    print(isSolveable(equation(10, [1, 9])))
    print(isSolveable(equation(10, [1, 8])))
    print(isSolveable(equation(10, [1, 10])))
    print(isSolveable(equation(10, [3, 3, 1])))

File: src/go/day7/puzzle1.go
package main

func isSolveable(eq equation) bool {
	if len(eq.terms) == 1 {
		return eq.target == eq.terms[0]
	}
	last := eq.terms[len(eq.terms)-1]
	// divide target by last
	if eq.target%last == 0 && isSolveable(equation{eq.target / last, eq.terms[:len(eq.terms)-1]}) {
		return true
	}

	// subtract last
	if eq.target > last && isSolveable(equation{eq.target - last, eq.terms[:len(eq.terms)-1]}) {
		return true
	}

	return false
}
